packages/malloc/src/pool.ts

```diff
@@ -10,42 +10,108 @@ import {
 } from "./api";
 import { wrap } from "./wrap";
 
+const POINTER_TOP_OFFSET = 0;
+const POINTER_END_OFFSET = POINTER_TOP_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+const POINTER_FREE_OFFSET = POINTER_END_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+const POINTER_USED_OFFSET = POINTER_FREE_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+
+const DATA_SPACE_OFFSET = POINTER_USED_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+
+
 export class MemPool implements IMemPool {
     buf: ArrayBuffer;
-    protected top: number;
-    protected start: number;
-    protected end: number;
+
+    /**
+     * Start is the anchor index inside the arraybuffer, so we can't save it inside the arraybuffer itself.
+     * If you pass the ArrayBuffer to other consumers they must use the same start value
+    */
+    protected readonly start: number;
+
     protected doCompact: boolean;
     protected doSplit: boolean;
     protected minSplit: number;
-    protected _free: MemBlock | null;
-    protected _used: MemBlock | null;
-
     protected u8: Uint8Array;
+    protected dataView: DataView;
```

Comment:
I'd favor renaming this `state` and using an `Uint32Array` view of that state info mem region. see below...